bundles/org.openhab.binding.enphase/src/main/java/org/openhab/binding/enphase/internal/discovery/EnvoyDiscoveryParticipant.java

```diff
@@ -0,0 +1,136 @@
+import org.openhab.core.config.discovery.mdns.MDNSDiscoveryParticipant;
+import org.openhab.core.thing.ThingTypeUID;
+import org.openhab.core.thing.ThingUID;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * MDNS discovery participant for discovering Envoy gateways.
+ * This service also keeps track of any discovered Envoys host name to provide this information for existing Envoy
+ * bridges
+ * so the bridge cat get the host name/ip address if that is unknown.
+ *
+ * @author Thomas Hentschel - Initial contribution
+ * @author Hilbrand Bouwkamp - Initial contribution
+ */
+@Component(service = { EnvoyHostAddressCache.class, MDNSDiscoveryParticipant.class })
+@NonNullByDefault
+public class EnvoyDiscoveryParticipant implements MDNSDiscoveryParticipant, EnvoyHostAddressCache {
+    private static final String ENVOY_MDNS_ID = "envoy";
+
+    private final Logger logger = LoggerFactory.getLogger(EnvoyDiscoveryParticipant.class);
+
+    private final Map<String, @Nullable String> lastKnownHostAddresses = new ConcurrentHashMap<>();
+
+    @Override
+    public Set<ThingTypeUID> getSupportedThingTypeUIDs() {
+        return Collections.singleton(EnphaseBindingConstants.THING_TYPE_ENPHASE_ENVOY);
+    }
+
+    @Override
+    public String getServiceType() {
+        return "_enphase-envoy._tcp.local.";
+    }
+
+    @Override
+    public @Nullable DiscoveryResult createResult(final ServiceInfo info) {
+        final String id = info.getName();
+
+        logger.debug("id found: {} with type: {}", id, info.getType());
+
+        if (!id.contains(ENVOY_MDNS_ID)) {
+            return null;
+        }
+
+        if (info.getInet4Addresses().length == 0 || info.getInet4Addresses()[0] == null) {
+            return null;
+        }
+
+        final ThingUID uid = getThingUID(info);
+
+        if (uid == null) {
+            return null;
+        }
+
+        final Inet4Address hostname = info.getInet4Addresses()[0];
+        final String serialNumber = info.getPropertyString(DISCOVERY_SERIAL);
+
+        if (serialNumber == null) {
+            logger.debug("No serial number found in data for discovered Envoy {}: {}", id, info);
+            return null;
+        }
+        final String version = info.getPropertyString(DISCOVERY_VERSION);
+        final String hostAddress = hostname == null ? "" : hostname.getHostAddress();
+
+        lastKnownHostAddresses.put(serialNumber, hostAddress);
+        final Map<String, Object> properties = new HashMap<>(3);
+
+        properties.put(CONFIG_SERIAL_NUMBER, serialNumber);
+        properties.put(CONFIG_HOSTNAME, hostAddress);
+        properties.put(PROPERTY_VERSION, version);
+        return DiscoveryResultBuilder.create(uid).withProperties(properties)
+                .withRepresentationProperty(CONFIG_SERIAL_NUMBER)
+                .withLabel("Enphase Envoy " + defaultPassword(serialNumber)).build();
```

Comment:
If the serial number is invalid wouldn't the defaultPassword end up being empty here? You shouldn't make the label display an empty string in that case, maybe just display the serial number anyway?

Reply:
I would not expect the serial number to be invalid here as it's directly read from the data returned by the device. It would not be completely empty because the label is prefixed with `Enphase Envoy`.

Reply:
Fair enough, just though it was worth pointing out.